File: solutions/198. House Robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        # [2,7,9,3,1,100]
        # [2,7,11,10,12,111]
        
        # time O(N), space O(1)
        rob, not_rob = 0,0
        not_rob_prev = 0
        for num in nums:
            not_rob_prev = not_rob
            not_rob = rob
            rob = max(not_rob_prev+num, not_rob)
        return max(rob, not_rob)
            
#         # time O(N), space O(N)
#         if len(nums) == 1:
#             return nums[0]
#         elif len(nums) == 2:
#             return max(nums)
        
#         rob_gain = nums
#         rob_gain[2] += rob_gain[0]
        
#         for i in range(3, len(nums)):
#             rob_gain[i] += max(rob_gain[i-2], rob_gain[i-3])
            
#         return max(rob_gain[-1], rob_gain[-2])
        
        

File: solutions/198. House Robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        // i-th house, rob it -> $A, not rob it -> $B
        int rob = 0, notRob = 0, notRobPrev;
        for (const int num : nums) {
            notRobPrev = notRob;
            notRob = rob;
            rob = max(notRobPrev + num, notRob);
        }
        return max(rob, notRob);
        
        // nums = [1,2,3,1]
        // num = 1, R = 1, N = 0, ans = 1
        // num = 2, R = 2, N = 1, ans = 2
        // num = 3, R = 4, N = 2, ans = 4
        // num = 1, R = 3, N = 
    } 
};
